library manager: initial implementation

pjdobrowolski opened the pull request:
Library manager is part of loading external libraries responsible of
handling data and placing it in proper order and places.

After receiving IPC4 load library command
 - reads manifest
 - after module_id choses module to load and verify it (WIP)
 - loads manifest
 - allocs propper memory size
 - transfer module from IMR to HPSRAM (WIP)

Signed-off-by: Stelter, Jaroslaw <jaroslaw.stelter@intel.com>
Signed-off-by: Dobrowolski, PawelX <pawelx.dobrowolski@intel.com>

sofci commented:
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

pjdobrowolski commented:
test this please

From: SOF CI ***@***.***>
Sent: wtorek, 10 maja 2022 12:48
To: thesofproject/sof ***@***.***>
Cc: Dobrowolski, PawelX ***@***.***>; Author ***@***.***>
Subject: Re: [thesofproject/sof] library manager: initial implementation (PR #5796)


Can one of the admins verify this patch?

reply test this please to run this test once

—
Reply to this email directly, view it on GitHub<https://github.com/thesofproject/sof/pull/5796#issuecomment-1122227992>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AWLAD4ERTPMWLIWH3KENKC3VJI5G7ANCNFSM5VRE6XNQ>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>
---------------------------------------------------------------------
Intel Technology Poland sp. z o.o.
ul. Slowackiego 173 | 80-298 Gdansk | Sad Rejonowy Gdansk Polnoc | VII Wydzial Gospodarczy Krajowego Rejestru Sadowego - KRS 101882 | NIP 957-07-52-316 | Kapital zakladowy 200.000 PLN.
Ta wiadomosc wraz z zalacznikami jest przeznaczona dla okreslonego adresata i moze zawierac informacje poufne. W razie przypadkowego otrzymania tej wiadomosci, prosimy o powiadomienie nadawcy oraz trwale jej usuniecie; jakiekolwiek przegladanie lub rozpowszechnianie jest zabronione.
This e-mail and any attachments may contain confidential material for the sole use of the intended recipient(s). If you are not the intended recipient, please contact the sender and delete all copies; any review or distribution by others is strictly prohibited.

pjdobrowolski commented:
test this please

gkbldcig commented:
Can one of the admins verify this patch?

dbaluta commented on src/include/sof/library_manager.h:
```diff
@@ -0,0 +1,115 @@
+									uint32_t instance_id,
+									uint32_t is_pages,
+									struct sof_man_module *mod);
+
```
This looks like an improper indentation.

dbaluta commented on src/include/sof/library_manager.h:
```diff
@@ -0,0 +1,115 @@
+/* SPDX-License-Identifier: BSD-3-Clause
```
Can you add some more details about how this works.

- how is the memory for loading the lib is managed. Who reserves it? 
   - it looks like the module is binary is sent by the Host AP and DSP parses and loads it. Is that right?
- lets try to make this as Intel agnostic as possible. It should be a generic functionality. So maybe avoid terms like IMR / HPSRAM.

pjdobrowolski replied:
Memory manager is being prepared and will be added soon as a part of a FW. I will consider memory names to be more universal.

marc-hb commented:
test this please

pjdobrowolski commented on src/include/sof/library_manager.h:
```diff
@@ -0,0 +1,115 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
```
This PR is IPC4 focused. Porting it to IPC3 might be possible. Only matter of priorities.

dbaluta replied:
Yes, but would be easier to address the matter in a generic way.

pjdobrowolski replied:
Correct. However there are major differences between IPC3/IPC4 and resolving them for generic matter exceeds this feature. It is matter of further discussion.

lgirdwood replied:
> Correct. However there are major differences between IPC3/IPC4 and resolving them for generic matter exceeds this feature. It is matter of further discussion.

The library manager core needs to export an IPC generic API. I would expect to also see an IPC4 client integration as part of this PR (that's separate file(s) from the library manager).

pjdobrowolski replied:
That is why it is Draft PR. I will publish all code in separate commits for easier review later.

lgirdwood commented:
@pjdobrowolski pls check your inbox, invite sent to autorun the CI.

lgirdwood reviewed (commented):
Some questions
1) How are we testing this today ? and how should we test in CI
2) Does this version support the dynamic linking or just the dynamic loading today ?

jxstelter commented:
> Some questions
> 
> 1. How are we testing this today ? and how should we test in CI
> 2. Does this version support the dynamic linking or just the dynamic loading today ?

marc-hb commented on src/library_manager/Kconfig:
```diff
@@ -0,0 +1,14 @@
+	default n
+	help
+	  This is support for dynamic modules loading.
+	  FDK developed modules could be used if enabled.
```
"FDK" must be expanded. It's not used anywhere in SOF or sof-docs yet.

marc-hb commented on src/library_manager/library_manager.c:
```diff
@@ -0,0 +1,458 @@
+		uint32_t instance_id, uint32_t core_id, void *mod_cfg);
+
+
+/* TODO: Manifest size should be defined somewhere - to check!!!*/
```
See comment at the top of `src/include/kernel/ext_manifest.h`
```
/*
 * Parts of this file are copied to ext_manifest.h in Linux kernel tree
 * and to the rimage tree. This duplication is OK because it is part of
 * a versioned ABI.
 */
```

marc-hb reviewed (commented):
> More changes are required to have something testable.

Maybe that's why it's still a draft?

pjdobrowolski commented on src/include/sof/library_manager.h:
```diff
@@ -0,0 +1,115 @@
+#include <rimage/sof/user/manifest.h>
+
+#define MAX_LIBS		16
+#define LIB_ID_SHIFT	12
```
> 1. How are we testing this today ? and how should we test in CI
For today feature is waiting for platform with IPC4. Until then PR will stay as a draft.
> 2. Does this version support the dynamic linking or just the dynamic loading today ?
Dynamic loading and linking is under develpment. For bench test libs are loaded under fixed address.

dbaluta replied:
@pjdobrowolski can you please add this information in the commit message. Would be good to know what it works with this commit.

pjdobrowolski commented on src/include/sof/library_manager.h:
```diff
@@ -0,0 +1,159 @@
+
+struct ext_library {
+	struct sof_man_fw_desc *desc[MAX_LIBS];
+};
```
What do you mean by shared? Idea is to be quite flexible in loading open and close source libs.

dbaluta replied:
Please use a common prefix for all macros and functions (e.g `LIB_MANAGER` for macros and `lib_manager` for functions.

lgirdwood replied:
Sorry, I mean will this structure be shared across cores ?

jxstelter replied:
Yes, since modules could be executed on any core, we should assume it could be accessed from different cores.

lgirdwood replied:
ok, we can it embed the struct coherent which allows the locking/cache ops in one place (and will be hooked into SPARSE too).

dbaluta commented on src/include/sof/library_manager.h:
```diff
@@ -0,0 +1,159 @@
+/*
+ * Library manager implementation for SOF is using codec adapter API for
+ * loadable native and external libraries. Depends of information received from
+ * it manages memory and locate accordingly to available space.
```
`codec adapter` is no longer used. Everything is called `module adapter` now. Also you are alternating `library_manager` and `lib_manager` in your code/file names.

I would suggest sticking to `lib_manager` in all places.

dbaluta commented on src/library_manager/library_manager.c:
```diff
@@ -0,0 +1,458 @@
+ */
+static int mm_map(uint32_t *address, uint32_t size, uint32_t attribs)
+{
+	uint32_t addr_align = HOST_PAGE_SIZE;
```
what is this doing? please provide a better name. 

E.g __lib_manager_mm_map().

Please use:
* `lib_manager_` prefix for all exported functions.
* `__lib_manager` or `_lib_manager` for all static functions.

pjdobrowolski replied:
currently managing memory in a dummy way. This function will be replaced if TLB support will be available.

ranj063 commented on src/library_manager/Kconfig:
```diff
@@ -0,0 +1,15 @@
+
+config LOADABLE_MODULES
+	bool "Loadable Modules Support"
+	default n
```
does this depend on IPC_MAJOR_4?

jxstelter replied:
Yes. We need to add dependency in Kconfig.

lgirdwood replied:
any update or ETA ?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+#define DMA_ELEM_SIZE		32
+#define PAGE_ALIGN			__aligned(HOST_PAGE_SIZE)
+
+
```
remove extra new line

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+}
+
+static uint8_t *lib_manager_get_instance_bss_address(uint32_t module_id, uint32_t instance_id,
+						struct sof_man_module *mod)
```
can you please fix alignment here and everywhere else please?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	struct sof_man_module *mod;
+	uint32_t module_id = IPC4_MOD_ID(ipc_config->id);
+	uint32_t entry_index = GET_MODULE_INDEX(module_id);
+	struct ipc4_base_module_cfg *base_cfg = ipc_specific_config;
```
can you please club all the struct definitions first followed by the rest for consistency with the rest of the code in the SOF repo

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+			base_cfg->is_pages, mod);
+	if (ret < 0) {
+		tr_err(&lib_manager_tr, "lib_manager_allocate_module() failed: %d", ret);
+		return 0;
```
why return 0? should be return ret and you should modify the return type to return the error

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	struct sof *my_sof = sof_get();
+
+	my_sof->ext_library = &loader_ext_lib;
+	return 0;
```
seems needless to return int here

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	struct comp_driver_info *new_drv_info;
+	struct comp_driver *drv;
+	struct sof_man_module *mod;
+	int ret = IPC4_FAILURE;
```
why do you need the init?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+			sizeof(struct comp_driver), align);
+	if (!drv) {
+		tr_err(&lib_manager_tr, "lib_manager_register_module(): alloc failed");
+		return -ENOMEM;
```
dont you need to free the new_drv_info in case of error here?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	ret = comp_register(new_drv_info);
+	if (ret < 0)
+		tr_err(&lib_manager_tr, "lib_manager_register_module() failed: %d", ret);
+	return ret;
```
and free all allocated memory in case of error here?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+{
+	/* allocate new buffer */
+	buffer->addr = (uintptr_t)rballoc_align(0, SOF_MEM_CAPS_DMA,
+						size, align);
```
one line?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+{
+	rfree((void *)buffer->addr);
+	memset(buffer, 0, sizeof(struct lib_manager_dma_buf));
+	return 0;
```
i think this function should be have void return type

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+{
+	uint32_t addr_align;
+	int ret = IPC4_FAILURE;
+	/* request DMA in the dir HMEM->LMEM */
```
missing new line

pjdobrowolski replied:
Thanks for in depth review.

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	/* get required address alignment for dma buffer */
+	ret = dma_get_attribute(dma_ext->dma.dmac, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+					&addr_align);
+	if (ret < 0)
```
please check all error paths and do a dma_put()

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+
+	ret = dma_sg_alloc(&config.elem_array, SOF_MEM_ZONE_RUNTIME,
+			   config.direction, elem_num, elem_size, elem_addr, 0);
+
```
remove newline

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+		return ret;
+
+	ret = dma_set_config(dma_ext->chan, &config);
+	if (ret < 0)
```
check and free previous dma alloc in case of error here and the rest of the error paths

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	struct sof_man_fw_desc *man_desc =
+			(struct sof_man_fw_desc *) (buff_ptr + SOF_MAN_ELF_TEXT_OFFSET);
+	uint32_t preload_size = man_desc->header.preload_page_count * HOST_PAGE_SIZE;
+	int ret = IPC4_FAILURE;
```
no need to init

pjdobrowolski replied:
imo removing ret, and replacing it with big multilevel 'if' is less readable. Is it really needed?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	/* Now update sof context with new library */
+	ret = lib_manager_update_sof_ctx((struct sof_man_fw_desc *)library_base_address, lib_id);
+
+	return 0;
```
return  lib_manager_update_sof_ctx((struct sof_man_fw_desc *)library_base_address, lib_id);?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+int lib_manager_load_library(uint32_t dma_id, uint32_t lib_id)
+{
+	uint32_t addr_align;
+	int ret = IPC4_FAILURE;
```
remove init

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+{
+	uint32_t addr_align;
+	int ret = IPC4_FAILURE;
+
```
remove new line

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	int ret = IPC4_FAILURE;
+
+	lib_manager_init();
+	struct lib_manager_dma_ext dma_ext;
```
move with the rest of the variables?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	lib_manager_init();
+	struct lib_manager_dma_ext dma_ext;
+
+	ret = lib_manager_dma_init(&dma_ext, dma_id);
```
dont you need to check the return value?

pjdobrowolski replied:
please check new flow

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	/*      Allocate/map destination memory                     */
+	/*      load library body into destination memory           */
+
+	ret = lib_manager_store_library(&dma_ext, lib_id, addr_align);
```
same here? return value check?

ranj063 commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+
+	ret = lib_manager_dma_deinit(&dma_ext);
+
+	return ret;
```
just return lib_manager_dma_deinit() in the previous line

ranj063 commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,120 @@
+		}
+	}
+	/* Update header and size, since message handle can be reused */
+	msg->header = header;
```
what is msg is NULL here?

lyakh replied:
looks like it has to be checked after line 55 above

dbaluta commented on src/include/sof/lib_manager.h:
```diff
@@ -0,0 +1,204 @@
+
+/*
+ * Library manager implementation for SOF is using codec adapter API for
+ * loadable native and external libraries. Depends of information received from
```
As already pointed out, there is no longer 'codec adapter' but we renamed this to 'module adapter' please fix the doc

dbaluta commented on src/include/sof/lib_manager.h:
```diff
@@ -0,0 +1,204 @@
+ *   | IPC4_GLB_LOAD_LIBRARY |                            |                          |         |
+ *   | --------------------> | lib_manager_load_library() |                          |         |
+ *   |                       | -------------------------> | Prepare HDA DMA transfer |         |
+ *   |                       |                            | -------                  |         |
```
Just go with `prepare DMA transfer` (not all dmas are HDA).

dbaluta commented on src/include/sof/lib_manager.h:
```diff
@@ -0,0 +1,204 @@
+
+#define GET_LIB_ID(module_id) (module_id >> LIB_ID_SHIFT)
+#define GET_MODULE_INDEX(module_id) (module_id & 0xFFF)
+
```
Use `LIB` and `lib` prefixes wherever possible. e.g `ipc_lib_msg` -> `lib_ipc_msg`.

dbaluta commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+	drv = (struct comp_driver *)rballoc_align(0, SOF_MEM_CAPS_RAM,
+			sizeof(struct comp_driver), align);
+	if (!drv) {
+		tr_err(&lib_manager_tr, "lib_manager_register_module(): alloc failed");
```
Here you will get a memory leak as you don't free new_drv_info.

dbaluta commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,454 @@
+/* TODO
+ * This is stub, that will be replaced when TLB support
+ * will be available.
+ */
```
Please explain what is this supposed to do.

pjdobrowolski replied:
This part of code is waiting for memory manager feature which is prepared by @rfredzim

pjdobrowolski replied:
Updated.

lgirdwood commented:
@pjdobrowolski is this really draft PR ? i.e untested and WIP ?

pjdobrowolski commented:
> @pjdobrowolski is this really draft PR ? i.e untested and WIP ?
@lgirdwood yes, indeed it is a draft PR. However it is tested on architectural level still WIP because of i.e. lack memory management features.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,449 @@
+/* TODO: Manifest size should be defined somewhere - to check!!!*/
+#define CAVS18_MAX_MANIFEST_SIZE  (38*1024)
+#define DMA_ELEM_SIZE		32
+#define PAGE_ALIGN			__aligned(HOST_PAGE_SIZE)
```
These macros need to go in the correct files or be reused in existing files..

pjdobrowolski replied:
I think that they should be in board defines. Upstream process is over I'll check if they can be there.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,449 @@
+	return dst->data;
+}
+
+static uint8_t *lib_manager_get_instance_bss_address(uint32_t module_id, uint32_t instance_id,
```
Do we have UT for any of these APIs ? It seems like these can be tested quite easily. No need immediately, but should be added at some point.

lgirdwood replied:
A UT would create a test module and attempt to load it on host using these APIs. i.e. the UT locally mallocs buffers and uses these as fake text, data and bss. The UT then send this fake structure to the APIs and validates the fake data is processed correctly. Fwiw, the UTs are also run under the valgrind memory checker.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,449 @@
+	struct sof_man_fw_desc *desc;
+	struct sof_man_module *mod;
+	struct ipc4_base_module_cfg *base_cfg = ipc_specific_config;
+	int ret = IPC4_FAILURE;
```
These APIs need to return POSIX errors upto the IPC layer which will return IPC3/IPC4 errors.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,449 @@
+	/* Wait till whole data acquired */
+	while (avail_bytes < size) {
+		/* get data sizes from DMA */
+		ret = dma_get_data_size(dma_ext->chan, &avail_bytes, &free_bytes);
```
Whats the plan for using the Zephyr DMA APIs here ?

lgirdwood replied:
@abonislawski @juimonen pls confirm if we can use native Zephyr DMA calls for all stages here.

lgirdwood reviewed (commented):
Any ETA for updates ?

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+		wait_delay(1);
+#else
+		Z_TIMEOUT_CYC(1);
+#endif
```
We are using the native Zephyr APIs for wait now so you should be able to use wait_delay() natively now.

lgirdwood replied:
oh, and CYC is clock specific, best to specify in a real time unit like uS or ms.

lgirdwood replied:
This is not resolved, native APIs are used now so we can use Zephyr native wait (no need for ifdef). Btw, pls dont use the Z_ API, IIUC this is for internal Zephyr only.

jxstelter commented on src/ipc/ipc4/handler.c:
```diff
@@ -508,13 +525,89 @@ static int ipc4_process_chain_dma(struct ipc4_message_request *ipc4)
 	type = ipc4->primary.r.type;
 
+	//Receiving sof style IPC formatted as ACE IPC.
+	//This does not fit any ACE IPC interpretation and will change in the future.
```
Not relevant to this PR. Should be removed or added in separate PR.

jxstelter commented on src/ipc/ipc4/handler.c:
```diff
@@ -508,13 +525,89 @@ static int ipc4_process_chain_dma(struct ipc4_message_request *ipc4)
 }
 
+static int ipc4_log_enable(struct ipc4_message_request *ipc4)
+{
```
Not relevant to this PR. Should be removed or added in separate PR.

jxstelter commented on src/ipc/ipc4/helper.c:
```diff
@@ -30,6 +30,11 @@
 #include <ipc4/module.h>
 #include <ipc4/error_status.h>
+#include <sof/lib_manager.h>
+/* TODO: this is wrong. This header should not be here, since lodable */
```
Comment seems to be not actual.

lgirdwood commented:
@pjdobrowolski ready for more review yet or still WIP ?

pjdobrowolski commented:
> @pjdobrowolski ready for more review yet or still WIP ?

I'm open for any review. It is still WIP though.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	config.dest_width = sizeof(uint32_t);
+	config.cyclic = 0;
+
+	ret = dma_sg_alloc(&config.elem_array, SOF_MEM_ZONE_RUNTIME,
```
IIUC we are ready to start using the native Zephyr APIs for DMA memory allocation/config. @abonislawski @juimonen pls confirm.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	return 0;
+}
+
+static int lib_manager_store_library(struct lib_manager_dma_ext *dma_ext, uint32_t lib_id,
```
best to have a some function comments on some of the bigger static functions so everyone can follow.

lgirdwood commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+		k_spinlock_key_t key;
+		/* No free element or list empty, create new handle  */
+		if (_ext_lib->lib_notif_count > LIB_MANAGER_LIB_NOTIX_MAX_COUNT)
+			return NULL;
```
we should emit some trace here to tell the user they have too many modules in the lib

lgirdwood commented on src/include/ipc/topology.h:
```diff
@@ -48,6 +48,7 @@ enum sof_comp_type {
 	SOF_COMP_DCBLOCK,
 	SOF_COMP_SMART_AMP,		/**< smart amplifier component */
 	SOF_COMP_MODULE_ADAPTER,		/**< module adapter */
+	SOF_COMP_DYNAMIC,		/**< dynamic lodable library */
```
Does FW need this ? i.e. wont we have a custom IPC to load library ? I dont think driver will use this. @plbossart @ujfalusi @ranj063 pls confirm if driver needs this ?

ujfalusi replied:
I believe we do.. it looks like we are already lagging behind as the kernel is missing `SOF_COMP_MODULE_ADAPTER` type.

lgirdwood replied:
> I believe we do.. it looks like we are already lagging behind as the kernel is missing `SOF_COMP_MODULE_ADAPTER` type.

What would the kernel use this for ?

pjdobrowolski replied:
Can we leave it as is?

pjdobrowolski replied:
That line of code fires up ABI accordance error. Can it be as is?

RanderWang replied:
please don't add more component type in sof_comp_type which is not supported by ipc4. Please check that there is no copier type.

RanderWang replied:
why we need a new module type to load module ? Loading module should be a function of base_fw, so we can do it by base module.

pjdobrowolski replied:
Library manager is responsible for loading closed and opensource module. Ipc specific config size is not known for base firmware but library manager must pass this configuration. That is why we need new type of load module to not affect legacy sof mechanizm. Look up [src/ipc/ipc4/helper.c](https://github.com/pjdobrowolski/sof_zephyr/blob/c238559dd7adb88354a9f46e4d6047d71f1c76b0/src/ipc/ipc4/helper.c#L105)

lgirdwood replied:
@ujfalusi @ranj063 fyi - you may need this on Linux driver. Please review/comment.

ranj063 replied:
@pjdobrowolski why do you need to know the IPC size at all? In my PR, I pass the pointer to the mailbox payload as is without the size info https://github.com/thesofproject/sof/pull/5994/commits/80a947da9850113cc13f79ad7742889db6d631ac

pjdobrowolski replied:
It works for sof but in legacy platforms payload size would change depends on what module you load.

jxstelter replied:
When you load built in module, the base FW exactly knows its configuration size since it has all module headers already. For loadable modules configuration is proprietary and base FW can not figure out how large is the config in IPC buffer. To pass it to the module instance we need to know the size or pass whole 4kB IPC buffer.

ranj063 replied:
@jxstelter what is the downside to passing the entire mailbox buffer like in my PR?

jxstelter replied:
There are two problems for loadable modules:
1. Memory used - currently module adapter allocates buffer for the configuration and stores it in struct processing_module->priv pointer. We would need to allocate 4kB here. This is waste of memory.
2. Entry point function for IADK loadable modules expects within its parameters pointer to config buffer + size of valid data. 
Already developed modules will not start without correct configuration size.

ranj063 replied:
> 1. Memory used - currently module adapter allocates buffer for the configuration and stores it in struct processing_module->priv pointer. We would need to allocate 4kB here. This is waste of memory.

@jxstelter have you seen my PR https://github.com/thesofproject/sof/pull/5994/commits/13cdc38099d5bedbeea5774128cb74080ba8f994? I do no allocate memory for the config but instead just pass the pointer to the IPC payload directly

jxstelter replied:
> > 1. Memory used - currently module adapter allocates buffer for the configuration and stores it in struct processing_module->priv pointer. We would need to allocate 4kB here. This is waste of memory.
> 
> @jxstelter have you seen my PR [13cdc38](https://github.com/thesofproject/sof/commit/13cdc38099d5bedbeea5774128cb74080ba8f994)? I do no allocate memory for the config but instead just pass the pointer to the IPC payload directly

OK, this resolves problem with additional allocation. However still we need pass configuration size to IADK modules.

ranj063 replied:
@jxstelter do you have an example of where this size of valid data coems from?

jxstelter replied:
It is provided in IPC header for ipc4_init_module_instance():
**module.extension.r.param_block_size**

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	struct sof_man_fw_desc *desc;
+	struct sof_man_module *mod;
+	struct ipc4_base_module_cfg *base_cfg = ipc_specific_config;
+	int ret = -ENOMEM;
```
superfluous initialisation

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+{
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	struct ext_library *_ext_lib = ext_lib_get();
+	char *buffptr = (char *)_ext_lib->desc[lib_id];
```
`uint8_t` instead of `char`

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	uint32_t align = 4;
+	uint32_t entry_index = LIB_MANAGER_GET_MODULE_INDEX(module_id);
+
+	new_drv_info = (struct comp_driver_info *)rballoc_align(0, SOF_MEM_CAPS_RAM,
```
no need for a type-cast: allocations return `void *`

lyakh replied:
sorry, what kind of compilation crash? This really shouldn't be needed. Check all other calls to this function in the code. Particularly makes no sense at all to type-cast it to `(void *)`. Where you type-cast it to an integer - that I can understand, that the compiler would complain about.

pjdobrowolski replied:
To be sure, do you mean to remove (void *) casts not (struct comp_driver_info *) as it is showed in github?

lyakh replied:
all of them

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+		return ret;
+	}
+
+	drv = (struct comp_driver *)rballoc_align(0, SOF_MEM_CAPS_RAM,
```
ditto

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+
+	if (!new_drv_info) {
+		tr_err(&lib_manager_tr, "lib_manager_register_module(): alloc failed");
+		return ret;
```
just `return -ENOMEM;` here and below and remove `ret` initialisation

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+
+	/* Fill the new_drv_info structure with already known parameters */
+	/* Check already registered components */
+	mod = (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(entry_index));
```
`uint8_t` is preferred over `char` for 1-byte types

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+
+	new_drv_info->drv = drv;
+
+	// Register new driver in the list
```
C-style comment

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	}
+
+	bzero((void *)buffer->addr, size);
+	dcache_writeback_region((void *)buffer->addr, size);
```
Please use `(__sparse_force void __sparce_cache *)` to avoid sparse warnings

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+static int lib_manager_dma_init(struct lib_manager_dma_ext *dma_ext, uint32_t dma_id)
+{
+	uint32_t addr_align;
+	int ret = -ENOMEM;
```
superfluous init

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	struct dma_sg_config config;
+	uint32_t elem_addr;
+	uint32_t elem_num, elem_size;
+	int ret = -ENOMEM;
```
ditto

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+
+	elem_addr = (uint32_t)dma_ext->dmabp.addr;
+
+	elem_size = sizeof(uint64_t) * DMA_ELEM_SIZE;
```
why are we counting element sizes in 64-bit units?

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	ret = dma_sg_alloc(&config.elem_array, SOF_MEM_ZONE_RUNTIME,
+			   config.direction, elem_num, elem_size, elem_addr, 0);
+
+	if (ret < 0)
```
don't we need to `dma_sg_free()` in all these failure cases?

pjdobrowolski replied:
Yes.

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+
+	dma_sg_free(&config.elem_array);
+
+	dcache_invalidate_region((void *)dma_ext->dmabp.addr, size);
```
why is this here? Have we just used DMA to copy data into that area? If so, them it's too late to invalidate caches here. If they really have to be invalidated, then before the DMA transfer. If you end up using this and moving it elsewhere, please add the sparse annotations as explained above

jxstelter replied:
We just used DMA to copy new data to the buffer, but before accessing them we need invalidate cache. Otherwise we could read old data from cache instead of the new one just transferred using DMA.

lyakh replied:
if we had data in cache for those addresses, while DMA was writing there the CPU could speculatively write caches back to RAM too. Therefore caches should be invalidated before letting DMA write to that memody

pjdobrowolski replied:
cache is used only once during loading then modules are moved from cache to physical memory.

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	uint32_t addr_align = HOST_PAGE_SIZE;
+	void *local_add;
+	/* allocate new buffer */
+	local_add = (void *)rballoc_align(0, SOF_MEM_CAPS_DMA,
```
allocation is already returning `void *`

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	struct sof_man_fw_desc *man_desc =
+			(struct sof_man_fw_desc *) (buff_ptr + SOF_MAN_ELF_TEXT_OFFSET);
+	uint32_t preload_size = man_desc->header.preload_page_count * HOST_PAGE_SIZE;
+	int ret = -ENOMEM;
```
superfluous init

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+
+	/* Copy data from temp_mft_buf to destination memory (pointed by library_base_address) */
+	memcpy_s((void *)library_base_address, CAVS18_MAX_MANIFEST_SIZE,
+			 (void *)buff_ptr, CAVS18_MAX_MANIFEST_SIZE);
```
BTW, why CAVS18?

pjdobrowolski replied:
Because library manager must support legacy platforms.

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+	dma_ext->dmabp.end_addr = dma_ext->dmabp.w_ptr + dma_ext->dmabp.size;
+	dma_ext->dmabp.addr = dma_ext->dmabp.r_ptr;
+
+	ret = lib_manager_load_data_from_host(dma_ext, (preload_size - CAVS18_MAX_MANIFEST_SIZE));
```
superfluous parentheses

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,455 @@
+int lib_manager_load_library(uint32_t dma_id, uint32_t lib_id)
+{
+	uint32_t addr_align;
+	int ret = -ENOMEM;
```
superfluous

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+
+	if (lib_notif) {
+		/* If list not empty search for free notification handle */
+		if (!list_is_empty(&lib_notif->list)) {
```
could you use `&&` to merge the two `if`s?

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+			list_for_item(list_elem, &lib_notif->list) {
+				msg_pool_elem = container_of(list_elem, struct ipc_lib_msg, list);
+				if (msg_pool_elem->msg != 0) {
+					if (list_is_empty(&msg_pool_elem->msg->list)) {
```
ditto

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+{
+	struct ipc_msg *msg = NULL;
+	struct ipc_lib_msg *msg_pool_elem;
+	struct ext_library *_ext_lib = ext_lib_get();
```
why begin with underscore?

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+			return NULL;
+		msg = ipc_msg_init(header, SRAM_OUTBOX_SIZE);
+		msg_pool_elem->msg = msg;
+		_ext_lib->lib_notif_count++;
```
doesn't the counter also have to be protected by the spin-lock?

pjdobrowolski replied:
Same case as below.

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+			k_spin_unlock(&_ext_lib->lock, key);
+		} else {
+			_ext_lib->lib_notif_pool = msg_pool_elem;
+		}
```
I must confess I'm confused by the special handling of the first message and by the `leave_one_handle` flag, could we have a more detailed explanation about them?

pjdobrowolski replied:
99% cases single notification buffer can be used by all modules but there is possible case when two modules on separate cores would race to the same buffer.

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+	}
+
+	/* Remove last handle - this should happen when there is no external libraries */
+	if ((!leave_one_handle) && (list_is_empty(&lib_notif->list))) {
```
superfluous parentheses on both sides

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+		ipc_msg_free(lib_notif->msg);
+		key = k_spin_lock(&_ext_lib->lock);
+		list_item_del(&lib_notif->list);
+		rfree(lib_notif);
```
`rfree()` doesn't need to be protected

pjdobrowolski replied:
k_spin_lock is race protection between notifications from more then one core.

pjdobrowolski replied:
It is overprotection right now until CI module test will be performed you can't say for 100%.

lyakh commented on src/include/ipc4/notification.h:
```diff
@@ -56,3 +56,28 @@ enum sof_ipc4_notification_type {
+
+		uint32_t _reserved_0 : 1;
+	} r;
+} __packed; __aligned(4);
```
the first semicolon seems to be wrong there

lyakh commented on src/include/ipc4/module.h:
```diff
@@ -372,6 +372,36 @@ struct ipc4_module_set_dx {
+			uint32_t _reserved_2		: 2;
+		} r;
+	} data;
+} __packed; __aligned(4);
```
like below - the first semicolon seems to be an error

lyakh commented on src/ipc/ipc-common.c:
```diff
@@ -204,8 +204,9 @@ void ipc_msg_send(struct ipc_msg *msg, void *data, bool high_priority)
-	if (msg->tx_size > 0 && msg->tx_size < SOF_IPC_MSG_MAX_SIZE) {
+	/* copy mailbox data to message if not already copied */
+	if ((msg->tx_size > 0 && msg->tx_size < SOF_IPC_MSG_MAX_SIZE) &&
+			(msg->tx_data != data)) {
```
all internal parentheses are superfluous

lyakh commented on src/ipc/ipc4/helper.c:
```diff
@@ -91,7 +95,15 @@ struct comp_dev *comp_new(struct sof_ipc_comp *comp)
+		spec.data = (unsigned char *)MAILBOX_HOSTBOX_BASE;
+		/* spec_size in IPC4 is in DW. Convert to bytes. */
+		spec.size = (comp->ext_data_length * 4);
+		dev = drv->ops.create(drv, &ipc_config, (void *)&spec);
```
don't think the typecast is needed

pjdobrowolski replied:
Is needed overwise there are compilation errors.

lyakh commented on src/ipc/ipc4/helper.c:
```diff
@@ -798,15 +813,39 @@ const struct comp_driver *ipc4_get_comp_drv(int module_id)
+		return NULL;
+#endif
+	}
+	// Check already registered components
```
comment style - here and below

RanderWang commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,683 @@
+						     struct sof_man_module *mod)
+{
+	uint32_t instance_bss_size =
+		 mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count;
```
is the r.length == bss_size * max_count ?  we can get bbs_size from module_config[module_cfg_offset].is_bytes and the bss_size depends on stream format

pjdobrowolski replied:
What about case when loadable library is being reconfigured after load? This mechanism is based on flow proved in closed source and works even if we change f.e. format using ipc set config.

RanderWang commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+			/* Search for not used message handle */
+			list_for_item(list_elem, &lib_notif->list) {
+				msg_pool_elem = container_of(list_elem, struct ipc_lib_msg, list);
+				if (msg_pool_elem->msg != 0) {
```
if (msg_pool_elem->msg != 0)  => if (msg_pool_elem->msg)

RanderWang commented on src/ipc/ipc4/handler.c:
```diff
@@ -481,6 +483,22 @@ static int ipc4_set_pipeline_state(struct ipc4_message_request *ipc4)
+	library.header.dat = ipc4->primary.dat;
+	library.data.dat = ipc4->extension.dat;
+
+	ret = lib_manager_load_library(library.header.r.dma_id, library.header.r.lib_id);
```
The ret value is a type of POSIX (.e.g -EINVAL) not IPC4 like IPC4_INVALID_RESOURCE_ID. Please refine the error code in lib_load_manager to IPC4 ( is it only for IPC4 ?) or convert POSIX error code to IPC4 one

pjdobrowolski replied:
This functionality is available only in closed source FW and firmware with IPC4. So that leads to why it is only IPC4. Code is split between FW and SoF module, that is why we have a mixture POSIX/IPC4. I've changed all ret from IPC4 to POSIX where I could, rest is not related to Zephyr.

RanderWang reviewed (commented):
please don't add more component type in sof_comp_type which is not supported by ipc4. Please check that there is no copier type.

pjdobrowolski commented on src/library_manager/library_manager.c:
```diff
@@ -0,0 +1,460 @@
+	struct lib_manager_dma_buf dmabp;
+};
+
+struct ext_library loader_ext_lib;
```
What would it change? Library manager can be run one and only once.

lyakh replied:
it changes the bloat. Everything, that can be static, should be static.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,681 @@
+				   struct sof_man_fw_desc *desc)
+{
+	struct ext_library *ext_lib = ext_lib_get();
+	uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	size_t load_offset = (size_t)((void *)ext_lib->desc[lib_id]);
+	void *va_base_text = (void *)mod->segment[SOF_MAN_SEGMENT_TEXT].v_base_addr;
+	void *src_txt = (void *)(mod->segment[SOF_MAN_SEGMENT_TEXT].file_offset + load_offset);
+	size_t st_text_size = mod->segment[SOF_MAN_SEGMENT_TEXT].flags.r.length;
+	void *va_base_rodata = (void *)mod->segment[SOF_MAN_SEGMENT_RODATA].v_base_addr;
+	void *src_rodata =
+		(void *)(mod->segment[SOF_MAN_SEGMENT_RODATA].file_offset + load_offset);
+	size_t st_rodata_size = mod->segment[SOF_MAN_SEGMENT_RODATA].flags.r.length;
```
These should all be treated as untrusted and validated before use.

pjdobrowolski replied:
Authentication, which is optional will be delivered in separate PR.

lgirdwood replied:
sorry, this is not authentication - but a basic security validation that the inputs are all in allowed ranges. i.e. module is not attempting to overwrite base FW text with it's own text by using an incorrect v_base_addr.

jxstelter replied:
I understand that this protection is done by memory manager in function **lib_manager_load_data_from_storage()**. The memory manager API function **sys_mm_drv_map_region()** tries to map virtual memory. If this memory is already used (mapped), the function will fail and we will not be able to use it.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,681 @@
+						uint32_t is_pages, struct sof_man_module *mod)
+{
+	int ret;
+	uint32_t bss_size =
+			(mod->segment[SOF_MAN_SEGMENT_BSS].flags.r.length / mod->instance_max_count)
+			 * CONFIG_MM_DRV_PAGE_SIZE;
+	uint8_t *va_base = lib_manager_get_instance_bss_address(module_id, instance_id, mod);
```
We also need to validate text, data, bss addresses and sizes. e.g. within bounds and cache/page aligned.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,681 @@
+	if (ret < 0)
+		return 0;
+
+	ret = lib_manager_allocate_module_instance(module_id, IPC4_INST_ID(ipc_config->id),
```
Can you get the instance ID in an IPC version generic manner ?

pjdobrowolski replied:
If it is going to support legacy modules it can't. Be assume that modules will be loadable without recompilation.

lgirdwood replied:
sorry - I mean this file needs to be IPC version generic. We could pass this ID in from IPC specific client code or could have a generic macro (that does something different depending on IPC version Kconfig) that gets the ID.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,681 @@
+{
+	struct sof_man_fw_desc *desc;
+	struct sof_man_module *mod;
+	uint32_t module_id = IPC4_MOD_ID(ipc_config->id);
```
ditto, needs to be IPC generic (this could be parsed in as an arg though).

pjdobrowolski replied:
See above.

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,681 @@
+{
+	struct ext_library *_ext_lib = ext_lib_get();
+
+	_ext_lib->desc[lib_id] = desc;
```
any locking here ?

pjdobrowolski replied:
don't need

pjdobrowolski commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,683 @@
+DECLARE_TR_CTX(lib_manager_tr, SOF_UUID(lib_manager_uuid), LOG_LEVEL_INFO);
+
+/* TODO: Manifest size should be defined somewhere - to check!!!*/
+#define CAVS18_MAX_MANIFEST_SIZE  (38 * 1024)
```
Where exactly? Rimage? Board definition? It is lagacy manifest size and it vary between CAVS15 and CAVS18.

lgirdwood replied:
The manifest is generated by rimage.

lgirdwood commented:
@pjdobrowolski can you rebase, this will unblock the CI. Thanks !

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,683 @@
+#include <sof/spinlock.h>
+#include <sof/lib_manager.h>
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/audio/module_adapter/module/iadk_modules.h>
```
@pjdobrowolski ping - what structures/macros etc do we depend on in iadk_modules.h ? These would be better defined in generic.h as a generic interface (which could also be included by iadk_modules.h).

lgirdwood commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,683 @@
+/* TODO: Manifest size should be defined somewhere - to check!!!*/
+#define CAVS18_MAX_MANIFEST_SIZE  (38 * 1024)
+#define DMA_ELEM_SIZE		32
+#define PAGE_ALIGN			__aligned(CONFIG_MM_DRV_PAGE_SIZE)
```
@jxstelter I now understand the Zephyr memory manager is merged, we should now use the Zephyr version of this macro used by memory manager.

lgirdwood reviewed (commented):
@pjdobrowolski any ETA when the review comments can be addressed now that memory manager is merged ? I think we are almost good now.

keqiaozhang commented:
SOFCI TEST

lgirdwood commented:
@pjdobrowolski are all review comments resolved now or just an intermediate push for testing ?

pjdobrowolski commented:
@lgirdwood Lets say that all review comments are resolved.

lgirdwood reviewed (commented):
@pjdobrowolski there are still quite a few unresolved comments and all afaict are simple ones to address. i.e. add trace, add comments, add sparse, move a lock. I can merge after this is done.

pjdobrowolski commented:
@lgirdwood I've done that, not all conversations were valid and caused some compilator errors. I've added trace as you suggested and rebased to upstream.

keqiaozhang commented:
SOFCI TEST

wszypelt commented:
@pjdobrowolski @lgirdwood Internal Intel CI System/merge/build and /codecheck are OK:
![image](https://user-images.githubusercontent.com/82375849/181231514-0a9ce354-20b7-471b-8bd2-bdabde77de50.png)
I dont know why this is not updated ;/
any idea?

lgirdwood commented:
@pjdobrowolski could you take a quick look at the CI, not sure why build would only fail for TGL (unless license server issue). @wszypelt fyi - may need to restart/check CI.

lgirdwood commented:
SOFCI TEST

lgirdwood commented:
@pjdobrowolski pls check CI again in a few hours, not expecting a TGL build failure but pls do fix if it comes back.

lgirdwood commented:
> @pjdobrowolski pls check CI again in a few hours, not expecting a TGL build failure but pls do fix if it comes back.

Please ignore me @pjdobrowolski - it seems CI is down for maintenance today. We can rerun tomorrow.

pjdobrowolski commented:
@lgirdwood I've to explain myself. Library manager might need slight refactor because of last changes to comp_drv. I'm checking it right now. So be patient I'm eager to finally marge it also.

pjdobrowolski commented:
@ranj063 I have a question about &md->cfg , from where comes configuration? Where is data from module init instance?

lgirdwood commented:
> @ranj063 I have a question about &md->cfg , from where comes configuration? Where is data from module init instance?

@pjdobrowolski Fwiw, @ranj063  is out. I will try and answer (although you may need to dig) - this should mostly come from topology (via IPC) although some data may also come from set_large_config() after the module has been created. No config should be hard coded.

pjdobrowolski commented:
@lgirdwood I agree, mod_cfg should come from ipc4, and be stored for specific module. If I understand, now we are using mod_cfg from mailbox which is not copied to any other temporary buffer and can be overwritten by upcoming ipc. Before https://github.com/thesofproject/sof/pull/5994 space for  mod_cfg was reserved for each module and that method was common for ipc3 and ipc4.

XiaoyunWu6666 commented:
SOFCI TEST

pjdobrowolski commented:
@lgirdwood I've resolved all issues. I think that it is ready to merge.

dbaluta commented:
Question for @pjdobrowolski  and @lgirdwood . How should we use this on i.MX?

Should we add an equivalent implementation for IPC3 or should we add support for IPC4 on Linux?

lgirdwood commented:
> Question for @pjdobrowolski and @lgirdwood . How should we use this on i.MX?
> 
> Should we add an equivalent implementation for IPC3 or should we add support for IPC4 on Linux?

This would be up to you and on your schedule. Intel are switching to IPC4 as its the same IPC used on Windows (i.e. less effort on Intel side). I've just checked with @ujfalusi and the kernel IPC4 is now all upstream except the trace.

dbaluta commented:
@lgirdwood thanks for the answer. We will also switch to IPC4 then on Linux because we don't want to be lagging on behind. Do you think IPC3 will be obsolete or we will keep supporting it in parallel?

dbaluta commented on src/include/sof/lib_manager.h:
```diff
@@ -0,0 +1,204 @@
+
+static SHARED_DATA struct ext_library {
+	struct k_spinlock lock;	/* locking mechanism */
+	struct sof_man_fw_desc *desc[LIB_MANAGER_MAX_LIBS];
```
better say here what data is protecting than `locking mechanism`. `locking mechanism` is too generic.

dbaluta commented on src/include/sof/lib_manager.h:
```diff
@@ -0,0 +1,204 @@
+static SHARED_DATA struct ext_library {
+	struct k_spinlock lock;	/* locking mechanism */
+	struct sof_man_fw_desc *desc[LIB_MANAGER_MAX_LIBS];
+	uint32_t mods_exec_load_cnt;
```
what exactly is `mods_exec_load_cnt`?

pjdobrowolski replied:
This feature is loading libraries which contains modules which shares some functions included in library. That counter is used for loading and unloading modules and shared funtions.

dbaluta commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,671 @@
+
+	ret = memcpy_s((void *)vma, size,
+		       (void *)s_addr, size);
+	if (ret < 0)
```
this doesn't need to be broken into multiple lines.

dbaluta commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,671 @@
+	 * Unmap bss memory and clear it.
+	 */
+	ret = sys_mm_drv_unmap_region(va_base, bss_size);
+	return ret;
```
you can directly `return` here. What does `unamp bss memory and clear it`? Does the unamp do the clear also?

pjdobrowolski replied:
Does not. Fixed and comment.

dbaluta commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,671 @@
+	struct sof_uuid *uid = (struct sof_uuid *)&mod->uuid[0];
+
+#if CONFIG_INTEL_MODULES
+	/* Declare new dynamic module with Module Adapter */
```
why is this under `CONFIG_INTEL_MODULES`. It doesn't really have any intel specific stuff.

pjdobrowolski replied:
Not exactly, library manager uses iadk_modules interface which is dependent on intel modules. That is why we used tath

dbaluta reviewed (commented):
@pjdobrowolski  I left some more comments. Also, please mark everything you fixed as 'Resolved'. 

I could see that there are some comments from @lyakh which are not resolved.

dbaluta commented on src/ipc/ipc-common.c:
```diff
@@ -190,8 +190,9 @@ void ipc_msg_send(struct ipc_msg *msg, void *data, bool high_priority)
+	    msg->tx_data != data) {
 		ret = memcpy_s(msg->tx_data, msg->tx_size, data, msg->tx_size);
 		assert(!ret);
 	}
```
Please put this change in another patch as it is a bugfix or an enhancement which is not really related directly to the current patch.

dbaluta commented on src/ipc/ipc4/handler.c:
```diff
@@ -32,6 +32,8 @@
 #include <user/trace.h>
 
```
can we please add changes from handler.c and helper.c in 1 or 2 separate patches. 

remember, one patch should do one single logical change. Unless there is an entire new feature added. 

You have added the lib manager new feature in 1 patch, but the usage of it, bugfixes, helpers must come in separate patches with proper commit description.

lgirdwood commented:
@jxstelter fyi since @pjdobrowolski is out.

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,671 @@
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
+
+	dcache_invalidate_region(va_base_rodata, st_rodata_size);
```
wouldn't the mm layer handle cache synchronisation? But in fact - this only synchronises cache on the current core. Couldn't other cores also access that memory? If yes - how do we handle their caches?

ujfalusi commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,671 @@
+		       (void *)s_addr, size);
+	if (ret < 0)
+		return ret;
+	dcache_writeback_region((void *)vma, size);
```
empty line before `dcache_writeback_region()`

ujfalusi commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,671 @@
+
+	/* Copy data from temp_mft_buf to destination memory (pointed by library_base_address) */
+	memcpy_s((void *)library_base_address, CAVS18_MAX_MANIFEST_SIZE,
+		 (void *)man_buffer, CAVS18_MAX_MANIFEST_SIZE);
```
Where this (CAVS18_MAX_MANIFEST_SIZE) is defined?

ujfalusi commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,671 @@
+			if (module_entry->type.lib_code) {
+				ret_tmp =
+				lib_manager_unload_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
+							  idx, mod, desc);
```
Is this going to end up in a dead-loop?

At least on recursion the `lib_manager_unload_module()` should be called with `module_entry` and _not_ with `mod` which is the functional part of the library?

On the first invocation of `lib_manager_unload_module()` from `lib_manager_free_module()` it is basically unload_library, or the free_module() itself is actually for the freeing of the whole library a given module is part of, no?

pjdobrowolski replied:
Dead-loop might happen, legacy code needs to be checked if there is similar case. Fixed that.

ujfalusi reviewed (commented):
@pjdobrowolski, do you have any library I can try to load with my Linux side PR (https://github.com/thesofproject/linux/pull/3826)?

pjdobrowolski commented:
> @pjdobrowolski, do you have any library I can try to load with my Linux side PR ([thesofproject/linux#3826](https://github.com/thesofproject/linux/pull/3826))?

You can build one yourself using https://github.com/thesofproject/converged-sof-modules

FYI: I came back from 2 week long sick leave and you guys flooded me with changes which should be requested almost 2 months ago. @dbaluta @ujfalusi @lgirdwood @lyakh I understand that next few days I will spend on resolving that requests. And I thought that sinusitis was bad...

ujfalusi commented on src/include/sof/lib_manager.h:
```diff
@@ -0,0 +1,204 @@
+
+static inline struct ext_library *ext_lib_get(void)
+{
+	return sof_get()->ext_library;
```
error: 'struct sof' has no member named 'ext_library'

pjdobrowolski replied:
In previous your comment you pointed to struct ext_library defined in this include.

ujfalusi replied:
First to make the PR compile:
```diff
diff --git a/app/overlays/tgl/ipc4_overlay.conf b/app/overlays/tgl/ipc4_overlay.conf
index b986e772f5f9..ca0086981921 100644
--- a/app/overlays/tgl/ipc4_overlay.conf
+++ b/app/overlays/tgl/ipc4_overlay.conf
@@ -11,3 +11,5 @@ CONFIG_PCM_CONVERTER_FORMAT_S16_C32_AND_S32_C32=y
 CONFIG_PCM_CONVERTER_FORMAT_S16_C32_AND_S24_C32=y
 CONFIG_PCM_CONVERTER_FORMAT_S16_C32_AND_S16_C32=y
 
+CONFIG_LIBRARY_MANAGER=y
+
diff --git a/src/Kconfig b/src/Kconfig
index a24d0fc402cb..a98564f16702 100644
--- a/src/Kconfig
+++ b/src/Kconfig
@@ -19,3 +19,5 @@ rsource "schedule/Kconfig"
 rsource "ipc/Kconfig"
 
 rsource "math/Kconfig"
+
+rsource "library_manager/Kconfig"
```
then compilation will fail:
```
sof/zephyr/../src/include/sof/lib_manager.h:88:1: warning: useless storage class specifier in empty declaration
   88 | };
      | ^
sof/zephyr/../src/include/sof/lib_manager.h: In function 'ext_lib_get':
sof/zephyr/../src/include/sof/lib_manager.h:95:18: error: 'struct sof' has no member named 'ext_library'
   95 |  return sof_get()->ext_library;
      |                  ^~
```

ujfalusi replied:
`struct sof` via `src/include/sof/sof.h` does not have a member named `ext_library`

pjdobrowolski replied:
This feature won't work at tgl. It is mtl and newer only. Additionally that feature needs zephyr native drivers, dma, ipc4. I'm not sure what are you aiming for. This feature is works since the beggining on developer integration branch.

pjdobrowolski replied:
@ujfalusi correct, there is lack of struct pointer. Fixed.

ujfalusi commented on src/include/sof/lib_manager.h:
```diff
@@ -0,0 +1,204 @@
+	struct list_item list;
+};
+
+static SHARED_DATA struct ext_library {
```
`static SHARED_DATA` is not needed, this is a declaration.

lyakh replied:
@pjdobrowolski you removed `SHARED_DATA` but you left `static` - please remove it too

pjdobrowolski replied:
done

pjdobrowolski commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,671 @@
+	int ret = sys_mm_drv_map_region(vma, POINTER_TO_UINT(NULL),
+					size, SYS_MM_MEM_PERM_RW | SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0) {
+		ret = -ENOMEM;
```
I'm afraid that it was requested in previous reviews. I have no intension in hiding anything.

lyakh replied:
if you mean this comment https://github.com/thesofproject/sof/pull/5796#discussion_r905753274 then it was for an earlier version, the code has changed since then, so, I agree, just `return ret` should be good here and in more cases below

dbaluta commented:
> FYI: I came back from 2 week long sick leave and you guys flooded me with changes which should be requested almost 2 months ago. @dbaluta @ujfalusi @lgirdwood @lyakh I understand that next few days I will spend on resolving that requests. And I thought that sinusitis was bad...

@pjdobrowolski don't worry you will get used with us ;). glad that you are better now.

lgirdwood commented:
rerun CI, license server
```
License checkout failed: Cannot connect to license server system.
 The license server manager (lmgrd) has not been started yet,
 the wrong port@host or license file is being used, or the
 port or hostname in the license file has been changed.
Feature:       XTENSA_XCC_TIE
Server name:   fmylic7005.fm.intel.com
License path:  84300@xtensa01p.elic.intel.com:/srv/home/jenkins/xcc/install/tools/RG-2017.8-linux/XtensaTools/Tools/lic/license.dat:
FLEXnet Licensing error:-15,570.  System Error: 115 "Operation now in progress"
For further information, refer to the FLEXnet Licensing documentation,
available at "www.macrovision.com".
make[3]: *** [CMakeFiles/sof.dir/build.make:1248: CMakeFiles/sof.dir/src/math/sqrt_int16.c.o] Error 2
make[3]: *** Waiting for unfinished jobs....
License checkout failed: No such feature exists.
```

pjdobrowolski commented:
@dbaluta Are there any guidelines like that for reviewers or just "hit and run"?

ujfalusi commented on src/include/sof/lib_manager.h:
```diff
@@ -0,0 +1,204 @@
+	struct list_item list;
+};
+
+static struct ext_library {
```
This is a struct declaration... `static` should not be there.

lgirdwood commented:
@pjdobrowolski can you force push again, I've just merged the build fix for IPC4 and newer CC. Thanks

lgirdwood commented:
SOFCI TEST

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+	struct ipc_msg *msg = NULL;
+	struct ipc_lib_msg *msg_pool_elem;
+	struct ext_library *ext_lib = ext_lib_get();
+	struct ipc_lib_msg *lib_notif = ext_lib->lib_notif_pool;
```
your `.lib_notif_pool` is somewhat... unusual. Usually you'd just have a list head in the higher level object (in `ext_lib` in this case) to which you then link elements as you allocate them. Instead you have a pointer to `lib_notif_pool` there and once you allocate the first element and attach it there, you then *link* further allocated elements to the first one. That makes the list head instance in that object both - a list head and a list element. `list_empty()` now means not 0 elements, but 1... All that can be made to work of course but just is a bit unusual and one has to remember about this when working with this code.

pjdobrowolski replied:
So what would you suggest. That might be refactored for sure but I think that main goal is to port that feature from legacy and integrate. Many mechanisms used here are borrowed from older firmware and for sure they can be improved. We can do that later if it isn't so wrong.

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,117 @@
+		if (list_is_empty(&msg_pool_elem->msg->list)) {
+			k_spinlock_key_t key;
+
+			ipc_msg_free(msg_pool_elem->msg);
```
in fact this should either go under the spinlock or maybe better go after removing from the list. As is you free memory in an object that is still linked to a globally accessible list with no protection

lyakh commented on src/include/ipc4/module.h:
```diff
@@ -372,6 +372,36 @@ struct ipc4_module_set_dx {
 		} r;
 	} extension;
 } __attribute__((packed, aligned(4)));
+struct ipc4_module_load_library {
```
an empty line between structures would be good

lyakh commented on src/ipc/ipc4/helper.c:
```diff
@@ -254,7 +256,6 @@ static struct comp_buffer *ipc4_create_buffer(struct comp_dev *src, struct comp_
 		tr_err(&ipc_tr, "failed to get base config for src %#x", dev_comp_id(src));
 		return NULL;
 	}
-
```
unrelated

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,119 @@
+		if (!msg_pool_elem)
+			return NULL;
+		msg = ipc_msg_init(header, SRAM_OUTBOX_SIZE);
+		if (!msg)
```
you have to free `msg_pool_elem` here

lgirdwood commented:
@lrudyX @wszypelt can you check CI, not expecting this to fail where it did. Looks like a copy issue ? Thanks.

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,659 @@
+	ret = lib_manager_load_data_from_storage(va_base_text, src_txt,
+						 st_text_size, SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0) {
+		ret = -ENOMEM;
```
as above - no need to replace the original error code

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,659 @@
+	ret = lib_manager_load_data_from_storage(va_base_rodata, src_rodata,
+						 st_rodata_size, SYS_MM_MEM_PERM_EXEC);
+	if (ret < 0) {
+		ret = -ENOMEM;
```
and here too

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,659 @@
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_text, st_text_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
```
`ret = ret_tmp`

lyakh commented on src/library_manager/lib_manager.c:
```diff
@@ -0,0 +1,659 @@
+
+	ret_tmp = sys_mm_drv_unmap_region(va_base_rodata, st_rodata_size);
+	if (ret_tmp < 0)
+		ret = -ENOMEM;
```
ditto

wszypelt commented:
@lgirdwood yea... 
TGL was disabled, I fixed it and ran the tests again.

pjdobrowolski commented:
rebased, resolved, ready to merge?

lgirdwood commented:
> rebased, resolved, ready to merge?

Good to merge as soon as CI finishes.

wszypelt commented:
SOFCI TEST

lyakh commented on src/library_manager/lib_notification.c:
```diff
@@ -0,0 +1,120 @@
+		msg = ipc_msg_init(header, SRAM_OUTBOX_SIZE);
+		if (!msg)
+			rfree(msg_pool_elem);
+			return NULL;
```
You need braces here

pjdobrowolski replied:
So reopen PR and will fix it.
